dockerfiles/Dockerfile-toolchain

```diff
@@ -39,26 +47,79 @@ RUN sudo git clone https://github.com/facebook/rocksdb -b v5.17.2 /rocksdb
 # This builds and installs just the static lib for us
 RUN cd /rocksdb && sudo make static_lib PORTABLE=1 -j$(nproc) && sudo cp librocksdb.a /usr/local/lib/librocksdb_coda.a && sudo rm -rf /rocksdb && sudo strip -S /usr/local/lib/librocksdb_coda.a
 
+# save commits for pinned packages, to know if they should be updated in CI
+RUN sudo git clone https://github.com/CodaProtocol/coda /coda
+
+# submodules
+RUN for pkg in async_kernel digestif graphql_ppx ocaml-extlib ppx_optcomp rpc_parallel ; do \
+      # remove leading "-" indicating uninitialized
+      cd /coda && sudo bash -c "git submodule status src/external/$pkg | awk '{print \$1}' | sed s/-// > ~opam/opam-repository/$pkg.commit" ; \
+    done
+
+# use log commit for ocaml-sodium, ocaml-rocksdb, coda_base58, which are not submodules
+RUN for pkg in ocaml-sodium ocaml-rocksdb coda_base58 ; do \
+      cd /coda && sudo bash -c "git log --format=oneline -n 1 src/external/$pkg | awk '{print \$1}' > ~opam/opam-repository/$pkg.commit" ; \
+    done
+
+# repo was only needed for submodule commits, remove it
+RUN sudo rm -rf /coda
+
 # OPAM dependencies
 RUN git -C /home/opam/opam-repository pull
 RUN opam update -y && opam upgrade -y
 
+
 # spacetime OCaml compiler
 RUN opam switch create $OCAML_VERSION+spacetime
 RUN eval $(opam env)
 
 # Install OPAM packages for spacetime compiler; use ulimited stack for camlp4 build
 ADD /src/opam.export .
-RUN ulimit -s unlimited && opam switch import opam.export ; rm opam.export
+RUN ulimit -s unlimited && opam switch import opam.export
```

Comment:
Based on my limited understanding: for dependency sharing in docker, I think that installing the opam switches should come first.